sdks/java/extensions/protobuf/src/main/java/org/apache/beam/sdk/extensions/protobuf/ProtoRow.java

```diff
@@ -0,0 +1,556 @@
+      }
+      return null;
+    }
+  }
+
+  /**
+   * Overlay handler for the Well Known Type "Wrapper". These wrappers make it possible to have
+   * nullable primitives.
+   */
+  static class WrapperOverlay<ValueT> implements FieldOverlay<ValueT> {
+    private Descriptors.FieldDescriptor fieldDescriptor;
+    private Descriptors.FieldDescriptor valueDescriptor;
+    private FieldOverlay value;
+
+    WrapperOverlay(ProtoSchema protoSchema, Descriptors.FieldDescriptor fieldDescriptor) {
+      this.fieldDescriptor = fieldDescriptor;
+      this.valueDescriptor = fieldDescriptor.getMessageType().findFieldByName("value");
+      this.value = protoSchema.createFieldLayer(valueDescriptor, false);
+    }
+
+    @Override
+    public ValueT get(Message message) {
+      if (message.hasField(fieldDescriptor)) {
+
+        Message wrapper = (Message) message.getField(fieldDescriptor);
+        return (ValueT) value.get(wrapper);
+      }
+      return null;
+    }
+
+    @Override
+    public ValueT convertGetObject(Object object) {
+      return (ValueT) object;
+    }
+
+    @Override
+    public void set(Message.Builder message, ValueT value) {
+      if (value != null) {
+        DynamicMessage.Builder builder =
+            DynamicMessage.newBuilder(fieldDescriptor.getMessageType());
+        this.value.set(builder, value);
+        message.setField(fieldDescriptor, builder.build());
+      }
+    }
+
+    @Override
+    public Object convertSetObject(Object value) {
+      return value;
+    }
+
+    @Override
+    public Schema.Field getSchemaField() {
+      return Schema.Field.of(
+          fieldDescriptor.getName(),
+          ProtoSchema.convertType(valueDescriptor.getType()).withNullable(true));
+    }
+  }
+
+  /**
+   * Overlay handler for the Well Known Type "Timestamp". This wrappers converts from a single Row
+   * DATETIME and a protobuf "Timestamp" messsage.
+   */
+  static class TimestampOverlay implements FieldOverlay<Instant> {
+    protected Schema.Field field;
+    private Descriptors.FieldDescriptor fieldDescriptor;
+    private Descriptors.FieldDescriptor secondsDescriptor;
+    private Descriptors.FieldDescriptor nanosDescriptor;
+
+    TimestampOverlay(Descriptors.FieldDescriptor fieldDescriptor) {
+      this.fieldDescriptor = fieldDescriptor;
+      this.secondsDescriptor = fieldDescriptor.getMessageType().findFieldByName("seconds");
+      this.nanosDescriptor = fieldDescriptor.getMessageType().findFieldByName("nanos");
+      this.field =
+          Schema.Field.of(fieldDescriptor.getName(), Schema.FieldType.DATETIME).withNullable(true);
+    }
+
+    @Override
+    public Instant get(Message message) {
+      if (message.hasField(fieldDescriptor)) {
+        Message wrapper = (Message) message.getField(fieldDescriptor);
+        return convertGetObject(wrapper);
+      }
+      return null;
+    }
+
+    @Override
+    public Instant convertGetObject(Object object) {
```

Comment:
this loses precision.

Reply:
Like to make this switchable via options, Instant will be enough for most Beam use-cases and will make it work nicely with Beam SQL.

Reply:
Same ticket: https://issues.apache.org/jira/browse/BEAM-7518